### PR TITLE
SSPCAB implementation

### PR DESCRIPTION
**Pull request by djdameln**
# Description

- This PR adds SSPCAB to model components and adds it as an optional extension to DRAEM.

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/components/layers/sspcab.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+"""Self-Supervised Predictive Convolutional Attention (SSPCAB) block for reconstruction-based models."""
```
Comment (samet-akcay) [inline anomalib/models/components/layers/sspcab.py:1]: Can you provide a link to the original paper? Perhaps to the official implementation as well?

File: anomalib/models/components/layers/sspcab.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        super().__init__()
+
+        out_channels = channels // reduction_ratio
+        self.fc1 = nn.Linear(channels, out_channels)
+        self.fc2 = nn.Linear(out_channels, channels)
+
+    def forward(self, inputs: Tensor) -> Tensor:
+        """Forward pass through the attention module."""
+        # average pool
+        avg_pooled = inputs.mean(dim=(2, 3))
```
Comment (samet-akcay) [inline anomalib/models/components/layers/sspcab.py:29]: for future reference or inform the reader, maybe we could comment here why `dim=(2,3)` is averaged

File: anomalib/models/components/layers/sspcab.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    """SSPCAB block.
+
+    Args:
+        channels (int): Number of input channels.
+        kernel_size (int): Size of the receptive fields of the masked convolution kernel.
+        dilation (int): Dilation factor of the masked convolution kernel.
+        reduction_ratio (int): Reduction ratio of the attention module.
+    """
+
+    def __init__(self, channels, kernel_size: int = 1, dilation: int = 1, reduction_ratio: int = 8):
```
Comment (samet-akcay) [inline anomalib/models/components/layers/sspcab.py:53]: `channels: int` ?

**Review by samet-akcay — COMMENTED** [review]
I like it! Only few minor points

File: anomalib/models/draem/lightning_model.py [hunk 4]
```diff
@@ -30,12 +31,40 @@ class Draem(AnomalyModule):
...
+        def get_activation(name: str) -> Callable:
+            """Retrieves the activations.
+
+            Args:
+                name (str): Identifier for the retrieved activations.
+            """
+
+            def hook(_, __, output: Tensor):
+                """Hook for retrieving the activations."""
+                self.sspcab_activations[name] = output.detach()
```
Comment (ashwinvaidya17) [inline anomalib/models/draem/lightning_model.py:62]: I don't get this. If we detach the outputs then will it be able to compute gradient? My understanding might be wrong here but shouldn't we only detach the input?

Comment (djdameln) [reply anomalib/models/draem/lightning_model.py]: yes, good catch. I've changed it

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks!

File: anomalib/models/draem/config.yaml [hunk 5]
```diff
@@ -24,6 +24,8 @@ model:
   name: draem
   anomaly_source_path: null # optional, e.g. ./datasets/dtd
   lr: 0.0001
+  enable_sspcab: false
+  sspcab_lambda: 0.1
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/models/draem/config.yaml:28]: This would be possible, but I think it would be more clear for the user to have an explicit parameter to enable/disable SSPCAB. Especially because we would have to add some logic to enable/disable the sspcab block in the architecture of the torch model based on the value of the lambda parameter. As a user I would not expect the value of the lambda parameter to affect the architecture of the model.

**Review by samet-akcay — APPROVED** [review]
